doc: lwm2m location assistance library API usage (#9379)
nrfconnect/sdk-nrf: doc/nrf/libraries/networking/lwm2m_location_assistance.rst

* Query of P-GPS predictions - The P-GPS predictions are queried from nRF Cloud and provided back to the device through the LwM2M server.
  The predictions are stored in the device flash and injected to the GNSS module when needed.

API usage
*********

Cell based location
===================

Cell based location can be based only on current cell or current and neighboring cells.
Getting the information about the current cell is achieved by using the Connectivity Monitor object (ID 4).
By enabling the kconfig option :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT`, the connectivity monitor is populated with data about current cell.

> Getting the information about the neighboring cells is achieved by using a collection of ECID-Signal Measurement Information object (ID 10256).
> These are populated by calling :c:func:`lwm2m_ncell_handler_register` to register the listener for the neighborhood measurements and scheduling a measurement by calling :c:func:`lwm2m_ncell_schedule_measurement`.

Reviewer: ```suggestion
To get information about the neighboring cells, use a collection of ECID Signal Measurement Information objects (ID 10256).
To populate the objects, call the :c:func:`lwm2m_ncell_handler_register` function to register the listener for the neighborhood measurements and :c:func:`lwm2m_ncell_schedule_measurement` to schedule a measurement.
```

Author: ECID-Signal Measurement Information is intended as the object is named that in the registry

> The Ground Fix Location object has only one resource that needs to be addressed before sending a location request, `report_back`.

Reviewer: ```suggestion
The ground fix location object needs to address the ``report_back`` resource before sending a location request.
```
Not so sure about the capitalization.

> The back reporting tells the server if it needs to send the acquired location back to the device or not.

Reviewer: ```suggestion
Back reporting tells the server whether it needs to send the acquired location back to the device.
```

> If the location is sent back to the device, the location is stored only in the Ground Fix Location object.

Reviewer: Check the capitalization.

Author: Capitalization is intended as they are referring to the names for the objects in LwM2M registry

> To finally send the location request for the cell based location, call :c:func:`location_assistance_ground_fix_request_send`.

Reviewer: ```suggestion
To send the location request for the cell-based location, call the :c:func:`location_assistance_ground_fix_request_send` function.
```

A-GPS location
==============

When using A-GPS location, the device requests A-GPS assistance data from the server.

> GNSS-module can be queried for the data it needs. Device can ask all the data in one go or split the request to reduce the memory usage.

Reviewer 2: ```suggestion
GPS module can be queried for the data it needs. The device can ask for all the data in one go or split the request to reduce the memory usage.
```

Reviewer 2: Generally, I wouldn't talk about GNSS when this is GPS-only.

Author: The module is called GNSS module

> GNSS-module can be queried for the data it needs. Device can ask all the data in one go or split the request to reduce the memory usage.

Reviewer: ```suggestion
You can query the GNSS module for the data needed.
A device can request for all data at once or split the request to reduce the memory usage.
```

> The request also contains information about the current cell the device is connected to and the information is similarly available on Connectivity Monitor object as in the cell based location.

Reviewer: ```suggestion
The request also contains information about the current cell the device is connected to and the information is similarly available on Connectivity Monitor object as in the cell-based location.
```

> When requesting for A-GPS assistance data, the device should first set the mask for the data it is requesting by calling :c:func:`location_assistance_agps_set_mask`.

Reviewer: ```suggestion
When requesting for A-GPS assistance data, the device must first set the mask for the data it is requesting by calling the :c:func:`location_assistance_agps_set_mask` function.
```

> After the mask has been set, :c:func:`location_assistance_agps_request_send` sends the request with all the necessary data to the server and responds with the A-GPS assistance data.

Reviewer: ```suggestion
When the mask has been set, the :c:func:`location_assistance_agps_request_send` function sends the request with all necessary data to the server and responds with the A-GPS assistance data.
```

> The assistance data is written to the A-GPS module automatically by the library.

Reviewer 2: ```suggestion
The assistance data is written to the GPS module automatically by the library.
```

Filtered A-GPS
--------------

> There is also an option to use filtered A-GPS. In it, the satellites below the given angle above the ground will be filtered out.
> The angle can be set with :c:func:`location_assist_agps_set_elevation_mask` to a degree between `[0, 90]`. Setting the degree to -1 disables the filtering and is used by default.

Reviewer: ```suggestion
There is also an option to use filtered A-GPS.
With this option, the satellites below the given angle above the ground are be filtered out.
You can set the the angle to a degree `[0 - 90]` using the :c:func:`location_assist_agps_set_elevation_mask` function. Setting the degree to `-1` disables filtering.
This is the default setting.
```

Author: I think this feels clearer:
Setting the degree to `-1` disables filtering, which is the default setting.

P-GPS location
==============

When using P-GPS location, the device requests predictions for the satellites for a near future.

> P-GPS doesn't use the information about current cell at all, it store the information about satellites and injects the data to GNSS-module when needed.

Reviewer 2: ```suggestion
P-GPS doesn't use the information about the current cell at all, it stores the information about satellites and injects the data to the GPS module when needed.
```

> P-GPS doesn't use the information about current cell at all, it store the information about satellites and injects the data to GNSS-module when needed.

Reviewer: ```suggestion
P-GPS does not use information about current cell at all.
It stores the information about satellites and injects the data to the GNSS module when needed.
```

> External flash is needed when using P-GPS as the predictions needs 2kB of memory for each prediction.

Reviewer: ```suggestion
When using P-GPS, external flash is necessary as each prediction needs 2 kB of memory.
```

> When requesting for P-GPS assistance data, the device can set the P-GPS resources if needed.

Reviewer: ```suggestion
When requesting for P-GPS assistance data, the device can set the P-GPS resources.
```

> If using default values in the resources, the request will be for predictions for 1 week of time. (42 predictions, 7 days, 4 hours between predictions)

Reviewer: ```suggestion
If default values are used in the resources, predictions are requested for one week (42 predictions, 7 days, 4 hours between predictions).
```

> After setting the resources, :c:func:`location_assistance_pgps_request_send` sends the request to the server.

Reviewer: ```suggestion
When the resources have been set, the :c:func:`location_assistance_pgps_request_send` function sends the request to the server.
```

Result codes and automatic resend
=================================

> The location assistance objects both have a resource called `result_code`. This resource contains information about the request handling in the server side.

Reviewer: ```suggestion
The location assistance objects have a resource called ``result_code``.
This resource contains information about the request handling in the server side.
```

> It can have 3 different values:

Reviewer: ```suggestion
It can have three different values:
```

* 0  - The request was handled succesfully.

> * -1 - There is a permanent error in the server that needs fixing. Library will reject further requests and the device should be rebooted after the issue has been resolved in the server.

Reviewer: ```suggestion
* -1 - A permanent error in the server needs fixing.
   The library will reject further requests and the device must be rebooted after the issue has been resolved in the server.
```

> * 1  - There is a temporary error in the server and device should try again after a while.

Reviewer: ```suggestion
* 1  - There is a temporary error in the server and the device needs to try again after a while.
```
Try to do what?

Author: * 1  - There is a temporary error in the server and the device needs to retry sending the request after a while.

The library has a resend handler for the temporary error code. It can be initialized with :c:func:`location_assistance_init_resend_handler`.

> It uses an exponential backoff for scheduling the resends. User can also decide to not use this approach and set own callbacks for the result code for better control of the application.

Reviewer 2: ```suggestion
It uses an exponential backoff for scheduling the resends. Users can also decide to not use this approach and set their own callbacks for the result code for better control of the library.
```

> The library has a resend handler for the temporary error code. It can be initialized with :c:func:`location_assistance_init_resend_handler`.
> It uses an exponential backoff for scheduling the resends. User can also decide to not use this approach and set own callbacks for the result code for better control of the application.

Reviewer: ```suggestion
The library has a resend handler for the temporary error code.
It can be initialized with the :c:func:`location_assistance_init_resend_handler` function.
It uses an exponential backoff for scheduling the resends.
You can also set your own callbacks for the result code for better control of the application.
```

Using A-GPS and P-GPS simultaneously
====================================

> A-GPS and P-GPS can be used simultaneously. One needs to note that only one request can be ongoing at a time for the object.

Reviewer: ```suggestion
A-GPS and P-GPS can be used simultaneously.
However, only one active request at a time for the object is allowed.
```

> The :c:func:`location_assistance_agps_set_mask`, :c:func:`location_assistance_agps_request_send` and :c:func:`location_assistance_pgps_request_send` will return `-EAGAIN` if there is a request ongoing.

Reviewer: ```suggestion
The functions :c:func:`location_assistance_agps_set_mask`, :c:func:`location_assistance_agps_request_send` and :c:func:`location_assistance_pgps_request_send` return ``-EAGAIN`` if there is an active request.
```

> In that case, the device should try again after the previous request has been handled.

Reviewer: ```suggestion
In that case, the device must try again after the previous request has been handled.
```
Try to do what?

Author: * 1  - There is a temporary error in the server and the device needs to retry sending the request after a while.

* Query of P-GPS predictions - The P-GPS predictions are queried from nRF Cloud and provided back to the device through the LwM2M server.
  The predictions are stored in the device flash and injected to the GNSS module when needed.

API usage
*********

This section describes API usage in different scenarios.

Cell-based location
===================

> Cell-based location uses only on current cell or current and neighboring cells.

Reviewer 3: ```suggestion
Cell-based location uses only current cell or current and neighboring cells.
```

To get information about the current cell, use the Connectivity Monitor object (ID 4).
When you enable the Kconfig option :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT`, the connectivity monitor is populated with data about the current cell.

To get information about the neighboring cells, use a collection of ECID-Signal Measurement Information objects (ID 10256).
To populate the objects, call the :c:func:`lwm2m_ncell_handler_register` function to register the listener for the neighborhood measurements and :c:func:`lwm2m_ncell_schedule_measurement` to schedule a measurement.

The Ground Fix Location object needs to address the ``report_back`` resource before sending a location request.
Back reporting tells the server whether it needs to send the acquired location back to the device.
If the location is sent back to the device, the location is stored only in the Ground Fix Location object.

To send the location request for the cell-based location, call the :c:func:`location_assistance_ground_fix_request_send` function.

> A-GPS location

Reviewer 3: A-GPS doesn't itself provide location, so maybe we should talk about "A-GPS assistance" instead. If you change this, update also the first sentence below.

================

When using A-GPS assistance, the device requests A-GPS assistance data from the server.
You can query the GNSS module for the data needed.
A device can request for all data at once or split the request to reduce the memory usage.
The request also contains information about the current cell the device is connected to and the information is similarly available on Connectivity Monitor object as in the cell-based location.

When requesting for A-GPS assistance data, the device must first set the mask for the data it is requesting by calling the :c:func:`location_assistance_agps_set_mask` function.
When the mask has been set, the :c:func:`location_assistance_agps_request_send` function sends the request with all necessary data to the server and responds with the A-GPS assistance data.
The assistance data is written to the GNSS module automatically by the library.

Filtered A-GPS
--------------

> There is also an option to use filtered A-GPS.

Reviewer: ```suggestion
```
I think I already suggested leaving this out altogether. It does not add any value. Just get straight to the point.

> With filtered A-GPS, the satellites below the given angle above the ground are be filtered out.

Reviewer 3: ```suggestion
With filtered A-GPS, the satellites below the given angle above the ground are filtered out.
```

You can set the the angle to a degree `[0 - 90]` using the :c:func:`location_assist_agps_set_elevation_mask` function.
Setting the degree to `-1` disables filtering, which is the default setting.

> P-GPS location

Reviewer 3: Same here, I would use for example "P-GPS assistance" instead.

==============

When using P-GPS location, the device requests predictions for the satellites for a near future.
P-GPS does not use information about current cell at all.
It stores the information about satellites and injects the data to the GNSS module when needed.
When using P-GPS, external flash is necessary as each prediction needs 2 kB of memory.

> When requesting for P-GPS assistance data, the device can set the P-GPS resources.

Reviewer 3: Does a P-GPS request with default values work properly? The request specifies the date and time for the first prediction, so I don't know what happens if you don't set that.

Author: Yes it does. If you leave the starting day and starting time as 0, nRFCloud will treat them as current day and time.

If default values are used in the resources, predictions are requested for one week (42 predictions, 7 days, 4 hours between predictions).
When the resources have been set, the :c:func:`location_assistance_pgps_request_send` function sends the request to the server.

Result codes and automatic resend
=================================

The location assistance objects have a resource called ``result_code``.
This resource contains information about the request handling in the server side.
It can have three different values:

> * 0  - The request was handled successfully.
> * -1 - A permanent error in the server needs fixing.
>   The library will reject further requests and the device must be rebooted after the issue has been resolved in the server.
> * 1  - Due to a temporary error in the server, the device needs to retry sending the request after a while.

Reviewer: ```suggestion
* ``0``  - The request was handled successfully.
* ``-1`` - A permanent error in the server needs fixing.
  The library will reject further requests and the device must be rebooted after the issue has been resolved in the server.
* ``1``  - Due to a temporary error in the server, the device needs to retry sending the request after a while.
```

* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_PGPS` -  nRF Cloud provides P-GPS predictions and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL` -  nRF Cloud provides estimated location based on currently attached cell and its neighborhood.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS` - Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.

> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT` - Enable support for connectivity monitor utilities. Provides data about the current cell and network the device has connected to.

Reviewer: ```suggestion
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT` - Enable support for connectivity monitoring utilities.
   Provides data about the current cell and network the device has connected to.
```

* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_PGPS` -  nRF Cloud provides P-GPS predictions and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL` -  nRF Cloud provides estimated location based on currently attached cell and its neighborhood.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS` - Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.

> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT` - Enable support for connectivity monitoring utilities. Provides data about the current cell and network the device has connected to.

Reviewer: ```suggestion
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_CONN_MON_OBJ_SUPPORT` - Enable support for connectivity monitoring utilities.
  Provides data about the current cell and network the device has connected to.
```
Remember, each sentence on a separate line.
